presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java

```diff
@@ -96,6 +103,25 @@ private static OptionalLong reduce(OptionalLong first, OptionalLong second, Redu
         return OptionalLong.empty();
     }
 
+    public static <T> List<T> pickRandomSample(List<T> elements, int sampleSize)
+    {
+        requireNonNull(elements, "elements is null");
+        checkArgument(sampleSize >= 0, "sampleSize must be equal to or greater than zero");
+        if (elements.size() <= sampleSize) {
+            return elements;
+        }
+        if (sampleSize == 0) {
+            return ImmutableList.of();
+        }
+        ThreadLocalRandom random = ThreadLocalRandom.current();
+        if (sampleSize == 1) {
+            return ImmutableList.of(elements.get(random.nextInt(elements.size())));
+        }
+        List<T> copy = new ArrayList<>(elements);
+        shuffle(copy, random);
```

Comment:
if sample size is much lower than elements.size() then shuffling all elements sounds like an overkill

Reply:
So there are basically 2 algorithms for doing that. It is either shuffling + sublist, or random-cherry-picking + deduplication. Reshuffling takes O(array_size), but the algorithm is very cheap. Deduplication is O(sample_size), but it takes a hash-set maintaining that involves hashing + equals. The question is for what parameters what algorithm performs best? But the thing is that copying + shuffling an array of 100_000 elements takes ~0.1 ms. I don't think it make sense to go any further.

Reply:
You can go without de-duplication in O(sample_size)* time:

```
int n = l.size();
for (i = 0; i < sampleSize; i++) {
  int remaining = n - i;
  int pos = random(0, remaining); // ThreadLocalRandom
  swap(l, pos, n - i - 1);
}
return l.sublist(n - sampleSize, n);
```

*) of course you still need to make a copy of `elements`, so this is O(n). But copying should be way faster that shuffling all n elements.

I think this approach is statistically correct (no element is favored). Am I right?

Also, `Collections.shuffle` uses static `Random` instance, so has potential of contention.

Reply:
Actually i'm going to change the picking algorithm to make it deterministic, but still pseudo "random".